// @oruga-ui/oruga/dist/cjs/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var helpers = require('./helpers.js');
var config = require('./config.js');
var plugins$1 = require('./plugins-01bd04cc.js');
var Icon = require('./Icon-2540a413.js');
require('./FormElementMixin-bc2b3815.js');
var Input = require('./Input-bc558e4b.js');
var Autocomplete = require('./Autocomplete-299c127c.js');
var autocomplete = require('./autocomplete.js');
var Button = require('./Button-de8d0fc6.js');
var button = require('./button.js');
require('./InjectedChildMixin-f4307b88.js');
var carousel = require('./carousel.js');
require('./CheckRadioMixin-df88dd8e.js');
var Checkbox = require('./Checkbox-919a3162.js');
var checkbox = require('./checkbox.js');
var collapse = require('./collapse.js');
require('./MatchMediaMixin-c29cb02d.js');
require('./trapFocus-8381ef46.js');
var DropdownItem = require('./DropdownItem-14d90670.js');
var Select = require('./Select-47d48a49.js');
var Datepicker = require('./Datepicker-5882fcfd.js');
var datepicker = require('./datepicker.js');
var Timepicker = require('./Timepicker-9ac49884.js');
var datetimepicker = require('./datetimepicker.js');
var dropdown = require('./dropdown.js');
var Field = require('./Field-b0300e43.js');
var field = require('./field.js');
var icon = require('./icon.js');
var input = require('./input.js');
var inputitems = require('./inputitems.js');
require('./ssr-39c7e185.js');
var Loading = require('./Loading-36b28296.js');
require('./InstanceRegistry-70c57269.js');
var loading = require('./loading.js');
var notification = require('./notification.js');
var modal = require('./modal.js');
var menu = require('./menu.js');
var Pagination = require('./Pagination-11523258.js');
var pagination = require('./pagination.js');
var radio = require('./radio.js');
var select = require('./select.js');
var skeleton = require('./skeleton.js');
var sidebar = require('./sidebar.js');
var Tooltip = require('./Tooltip-2592db38.js');
var slider = require('./slider.js');
require('./SlotComponent-0a757062.js');
require('./TabbedChildMixin-08144384.js');
var steps = require('./steps.js');
var _switch = require('./switch.js');
var table = require('./table.js');
var tabs = require('./tabs.js');
var timepicker = require('./timepicker.js');
var tooltip = require('./tooltip.js');
var upload = require('./upload.js');

var plugins = /*#__PURE__*/Object.freeze({
    __proto__: null,
    Autocomplete: autocomplete.default,
    Button: button.default,
    Carousel: carousel.default,
    Checkbox: checkbox.default,
    Collapse: collapse.default,
    Datepicker: datepicker.default,
    Datetimepicker: datetimepicker.default,
    Dropdown: dropdown.default,
    Field: field.default,
    Icon: icon.default,
    Input: input.default,
    Inputitems: inputitems.default,
    Loading: loading.default,
    Modal: modal.default,
    Menu: menu.default,
    Notification: notification.default,
    NotificationNotice: notification.default,
    Pagination: pagination.default,
    Radio: radio.default,
    Select: select.default,
    Skeleton: skeleton.default,
    Sidebar: sidebar.default,
    Slider: slider.default,
    Steps: steps.default,
    Switch: _switch.default,
    Table: table.default,
    Tabs: tabs.default,
    Timepicker: timepicker.default,
    Tooltip: tooltip.default,
    Upload: upload.default
});

const Oruga = {
  install(Vue, options = {}) {
    config.setVueInstance(Vue); // Options

    const defaultConfig = config.getOptions();
    config.setOptions(helpers.merge(defaultConfig, options, true)); // Components

    for (const componentKey in plugins) {
      plugins$1.registerPlugin(Vue, plugins[componentKey]);
    } // Config component


    plugins$1.registerComponentProgrammatic(Vue, 'config', config.Programmatic);
  }

};
plugins$1.use(Oruga);

exports.Config = config.Plugin;
exports.ConfigProgrammatic = config.Programmatic;
exports.OIcon = Icon.__vue_component__;
exports.OInput = Input.__vue_component__;
exports.OAutocomplete = Autocomplete.__vue_component__;
exports.Autocomplete = autocomplete.default;
exports.OButton = Button.__vue_component__;
exports.Button = button.default;
exports.Carousel = carousel.default;
exports.OCarousel = carousel.OCarousel;
exports.OCarouselItem = carousel.OCarouselItem;
exports.OCheckbox = Checkbox.__vue_component__;
exports.Checkbox = checkbox.default;
exports.Collapse = collapse.default;
exports.OCollapse = collapse.OCollapse;
exports.ODropdown = DropdownItem.__vue_component__;
exports.ODropdownItem = DropdownItem.__vue_component__$1;
exports.OSelect = Select.__vue_component__;
exports.ODatepicker = Datepicker.__vue_component__;
exports.Datepicker = datepicker.default;
exports.BTimepicker = Timepicker.__vue_component__;
exports.BDatetimepicker = datetimepicker.BDatetimepicker;
exports.Datetimepicker = datetimepicker.default;
exports.Dropdown = dropdown.default;
exports.OField = Field.__vue_component__;
exports.Field = field.default;
exports.Icon = icon.default;
exports.Input = input.default;
exports.Inputitems = inputitems.default;
exports.OInputitems = inputitems.OInputitems;
exports.OLoading = Loading.__vue_component__;
exports.Loading = loading.default;
exports.LoadingProgrammatic = loading.LoadingProgrammatic;
exports.Notification = notification.default;
exports.NotificationNotice = notification.default;
exports.NotificationProgrammatic = notification.NotificationProgrammatic;
exports.ONotification = notification.ONotification;
exports.Modal = modal.default;
exports.ModalProgrammatic = modal.ModalProgrammatic;
exports.OModal = modal.OModal;
exports.Menu = menu.default;
exports.OMenu = menu.OMenu;
exports.OMenuItem = menu.OMenuItem;
exports.OMenuList = menu.OMenuList;
exports.OPagination = Pagination.__vue_component__;
exports.OPaginationButton = Pagination.__vue_component__$1;
exports.Pagination = pagination.default;
exports.ORadio = radio.ORadio;
exports.Radio = radio.default;
exports.Select = select.default;
exports.OSkeleton = skeleton.OSkeleton;
exports.Skeleton = skeleton.default;
exports.OSidebar = sidebar.OSidebar;
exports.Sidebar = sidebar.default;
exports.OTooltip = Tooltip.__vue_component__;
exports.OSlider = slider.OSlider;
exports.OSliderTick = slider.OSliderTick;
exports.Slider = slider.default;
exports.OStepItem = steps.OStepItem;
exports.OSteps = steps.OSteps;
exports.Steps = steps.default;
exports.OSwitch = _switch.OSwitch;
exports.Switch = _switch.default;
exports.OTable = table.OTable;
exports.OTableColumn = table.OTableColumn;
exports.Table = table.default;
exports.OTabItem = tabs.OTabItem;
exports.OTabs = tabs.OTabs;
exports.Tabs = tabs.default;
exports.Timepicker = timepicker.default;
exports.Tooltip = tooltip.default;
exports.OUpload = upload.OUpload;
exports.Upload = upload.default;
exports.default = Oruga;
