// @oruga-ui/oruga-next/dist/esm/Field.vue_vue_type_script_setup_true_lang-OMkZjnpI.mjs
/*! Oruga v0.10.5 | MIT License | github.com/oruga-ui/oruga */
import { defineComponent, useId, ref, watch, useSlots, computed, useTemplateRef, resolveComponent, createElementBlock, openBlock, normalizeClass, unref, renderSlot, createBlock, createCommentVNode, createTextVNode, toDisplayString, Fragment, renderList, resolveDynamicComponent, createSlots, withCtx, createElementVNode } from "vue";
import { g as getDefault } from "./config-B-gWlaT6.mjs";
import { isVNodeEmpty } from "./helpers.mjs";
import { d as defineClasses } from "./defineClasses-BApkclKh.mjs";
import { u as useMatchMedia } from "./useMatchMedia-DbB9F3o7.mjs";
import { i as injectField, p as provideField } from "./fieldInjection-Bs3TBpgZ.mjs";
const _hoisted_1 = ["id", "for"];
const _hoisted_2 = ["id", "for"];
const _sfc_main = /* @__PURE__ */ defineComponent({
  ...{
    isOruga: true,
    name: "OField",
    configField: "field"
  },
  __name: "Field",
  props: {
    override: { type: Boolean, default: void 0 },
    variant: { default: void 0 },
    label: { default: void 0 },
    labelSize: { default: () => getDefault("field.labelsize") },
    labelFor: { default: void 0 },
    labelId: { default: () => useId() },
    message: { default: void 0 },
    messageTag: { default: () => getDefault("field.messageTag", "p") },
    messageId: { default: () => useId() },
    grouped: { type: Boolean, default: false },
    addons: { type: Boolean, default: false },
    multiline: { type: Boolean, default: false },
    horizontal: { type: Boolean, default: false },
    mobileBreakpoint: { default: () => getDefault("field.mobileBreakpoint") },
    rootClass: {},
    mobileClass: {},
    focusedClass: {},
    filledClass: {},
    bodyClass: {},
    groupedClass: {},
    addonsClass: {},
    multilineClass: {},
    horizontalClass: {},
    horizontalLabelClass: {},
    horizontalBodyClass: {},
    labelClass: {},
    labelSizeClass: {},
    labelVariantClass: {},
    messageClass: {},
    messageVariantClass: {}
  },
  setup(__props) {
    const props = __props;
    const { isMobile } = useMatchMedia(props.mobileBreakpoint);
    const inputId = ref(props.labelFor);
    watch(
      () => props.labelFor,
      (v) => inputId.value = v
    );
    const fieldVariant = ref(props.variant);
    watch(
      () => props.variant,
      (v) => fieldVariant.value = v
    );
    const fieldMessage = ref(props.message);
    watch(
      () => props.message,
      (v) => fieldMessage.value = v
    );
    watch(fieldMessage, (value) => {
      if (parentField.value && parentField.value.hasInnerField) {
        if (!parentField.value.variant)
          parentField.value.setVariant(fieldVariant.value);
        if (!parentField.value.message) parentField.value.setMessage(value);
      }
    });
    const isFocused = ref(false);
    const isFilled = ref(false);
    const hasInnerField = ref(false);
    const { parentField } = injectField();
    if (parentField.value) parentField.value.addInnerField();
    const slots = useSlots();
    const hasLabel = computed(() => props.label || !!slots.label);
    const hasMessage = computed(() => !!fieldMessage.value || !!slots.message);
    const hasBody = computed(
      () => props.grouped || props.multiline || hasInnerField.value || hasAddons.value
    );
    const hasAddons = computed(
      () => props.addons && !props.horizontal && !!slots.default
    );
    function getInnerContent(vnode) {
      const slot = vnode();
      return slot.length === 1 && Array.isArray(slot[0].children) ? slot[0].children : slot;
    }
    const rootRef = useTemplateRef("rootElement");
    function addInnerField() {
      hasInnerField.value = true;
    }
    function setFocus(value) {
      isFocused.value = value;
    }
    function setFilled(value) {
      isFilled.value = value;
    }
    function setVariant(value) {
      fieldVariant.value = value;
    }
    function setMessage(value) {
      fieldMessage.value = value;
    }
    function setInputId(value) {
      inputId.value = value;
    }
    const inputAttrs = computed(() => ({
      "aria-labelledby": props.labelId,
      ...fieldVariant.value === "error" ? { "aria-errormessage": props.messageId } : { "aria-describedby": props.messageId }
    }));
    const provideData = computed(() => ({
      $el: rootRef.value,
      props,
      hasInnerField: hasInnerField.value,
      variant: fieldVariant.value,
      message: fieldMessage.value,
      labelId: props.labelId,
      inputAttrs: inputAttrs.value,
      addInnerField,
      setInputId,
      setFocus,
      setFilled,
      setVariant,
      setMessage
    }));
    provideField(provideData);
    const rootClasses = defineClasses(
      ["rootClass", "o-field"],
      [
        "horizontalClass",
        "o-field--horizontal",
        null,
        computed(() => !!props.horizontal)
      ],
      ["mobileClass", "o-field--mobile", null, isMobile],
      ["focusedClass", "o-field--focused", null, isFocused],
      ["filledClass", "o-field--filled", null, isFilled]
    );
    const labelClasses = defineClasses(
      ["labelClass", "o-field__label"],
      [
        "labelSizeClass",
        "o-field__label-",
        computed(() => props.labelSize),
        computed(() => !!props.labelSize)
      ],
      [
        "labelVariantClass",
        "o-field__label-",
        fieldVariant,
        computed(() => !!fieldVariant.value)
      ]
    );
    const horizontalLabelClasses = defineClasses([
      "horizontalLabelClass",
      "o-field__horizontal-label"
    ]);
    const horizontalBodyClasses = defineClasses([
      "horizontalBodyClass",
      "o-field__horizontal-body"
    ]);
    const bodyClasses = defineClasses(["bodyClass", "o-field__body"]);
    const innerBodyClasses = defineClasses(
      ["rootClass", "o-field"],
      ["groupedClass", "o-field--grouped", null, computed(() => props.grouped)],
      [
        "addonsClass",
        "o-field--addons",
        null,
        computed(() => !props.grouped && hasAddons.value)
      ],
      [
        "multilineClass",
        "o-field--multiline",
        null,
        computed(() => props.multiline)
      ]
    );
    const messageClasses = defineClasses(
      ["messageClass", "o-field__message"],
      [
        "messageVariantClass",
        "o-field__message-",
        fieldVariant,
        computed(() => !!fieldVariant.value)
      ]
    );
    return (_ctx, _cache) => {
      const _component_OField = resolveComponent("OField");
      return openBlock(), createElementBlock("div", {
        ref: "rootElement",
        "data-oruga": "field",
        class: normalizeClass(unref(rootClasses))
      }, [
        _ctx.horizontal ? (openBlock(), createElementBlock("div", {
          key: 0,
          class: normalizeClass(unref(horizontalLabelClasses))
        }, [
          hasLabel.value ? (openBlock(), createElementBlock("label", {
            key: 0,
            id: _ctx.labelId,
            for: inputId.value,
            class: normalizeClass(unref(labelClasses))
          }, [
            renderSlot(_ctx.$slots, "label", { label: _ctx.label }, () => [
              createTextVNode(toDisplayString(_ctx.label), 1)
            ])
          ], 10, _hoisted_1)) : createCommentVNode("", true)
        ], 2)) : (openBlock(), createElementBlock(Fragment, { key: 1 }, [
          hasLabel.value ? (openBlock(), createElementBlock("label", {
            key: 0,
            id: _ctx.labelId,
            for: inputId.value,
            class: normalizeClass(unref(labelClasses))
          }, [
            renderSlot(_ctx.$slots, "label", { label: _ctx.label }, () => [
              createTextVNode(toDisplayString(_ctx.label), 1)
            ])
          ], 10, _hoisted_2)) : createCommentVNode("", true)
        ], 64)),
        _ctx.horizontal ? (openBlock(), createElementBlock("div", {
          key: 2,
          class: normalizeClass(unref(horizontalBodyClasses))
        }, [
          (openBlock(true), createElementBlock(Fragment, null, renderList(getInnerContent(_ctx.$slots.default), (element, index) => {
            return openBlock(), createElementBlock(Fragment, { key: index }, [
              unref(isVNodeEmpty)(element) ? (openBlock(), createBlock(resolveDynamicComponent(element), { key: 0 })) : (openBlock(), createBlock(_component_OField, {
                key: 1,
                variant: fieldVariant.value,
                addons: false,
                "label-id": _ctx.labelId,
                "message-id": _ctx.messageId,
                "message-tag": _ctx.messageTag,
                "message-class": _ctx.messageClass
              }, createSlots({
                default: withCtx(() => [
                  (openBlock(), createBlock(resolveDynamicComponent(element)))
                ]),
                _: 2
              }, [
                index === 0 ? {
                  name: "message",
                  fn: withCtx(() => [
                    renderSlot(_ctx.$slots, "message", { message: fieldMessage.value }, () => [
                      createTextVNode(toDisplayString(fieldMessage.value), 1)
                    ])
                  ]),
                  key: "0"
                } : void 0
              ]), 1032, ["variant", "label-id", "message-id", "message-tag", "message-class"]))
            ], 64);
          }), 128))
        ], 2)) : hasBody.value ? (openBlock(), createElementBlock("div", {
          key: 3,
          class: normalizeClass(unref(bodyClasses))
        }, [
          createElementVNode("div", {
            class: normalizeClass(unref(innerBodyClasses))
          }, [
            renderSlot(_ctx.$slots, "default")
          ], 2)
        ], 2)) : renderSlot(_ctx.$slots, "default", { key: 4 }),
        hasMessage.value && !_ctx.horizontal ? (openBlock(), createBlock(resolveDynamicComponent(_ctx.messageTag), {
          key: 5,
          id: _ctx.messageId,
          class: normalizeClass(unref(messageClasses))
        }, {
          default: withCtx(() => [
            renderSlot(_ctx.$slots, "message", { message: fieldMessage.value }, () => [
              createTextVNode(toDisplayString(fieldMessage.value), 1)
            ])
          ]),
          _: 3
        }, 8, ["id", "class"])) : createCommentVNode("", true)
      ], 2);
    };
  }
});
export {
  _sfc_main as _
};
