// @oruga-ui/oruga-next/dist/esm/Select.vue_vue_type_script_setup_true_lang-C8hU8ZRT.mjs
/*! Oruga v0.10.5 | MIT License | github.com/oruga-ui/oruga */
import { defineComponent, mergeModels, useTemplateRef, useModel, watch, computed, useAttrs, createElementBlock, openBlock, normalizeClass, unref, createBlock, createCommentVNode, withDirectives, createElementVNode, mergeProps, renderSlot, Fragment, createTextVNode, toDisplayString, renderList, vShow, vModelSelect, useId, nextTick } from "vue";
import { _ as _sfc_main$1 } from "./Icon.vue_vue_type_script_setup_true_lang-v4MwBFqk.mjs";
import { g as getDefault } from "./config-B-gWlaT6.mjs";
import { isTrueish, isDefined } from "./helpers.mjs";
import { d as defineClasses } from "./defineClasses-BApkclKh.mjs";
import { u as useInputHandler } from "./useInputHandler-B842u-_G.mjs";
import { n as normalizeOptions, i as isGroupOption } from "./useOptions-rS_XNoke.mjs";
import { u as useSequentialId } from "./useSequentialId-aw-jeTgs.mjs";
import { i as injectField } from "./fieldInjection-Bs3TBpgZ.mjs";
const _hoisted_1 = ["id", "autocomplete", "multiple", "size", "disabled", "required"];
const _hoisted_2 = {
  key: 0,
  value: "",
  disabled: "",
  hidden: ""
};
const _hoisted_3 = ["label", "value"];
const _hoisted_4 = ["value", "selected"];
const _hoisted_5 = ["value", "selected"];
const _sfc_main = /* @__PURE__ */ defineComponent({
  ...{
    isOruga: true,
    name: "OSelect",
    configField: "select",
    inheritAttrs: false
  },
  __name: "Select",
  props: /* @__PURE__ */ mergeModels({
    override: { type: Boolean, default: void 0 },
    modelValue: { default: void 0 },
    multiple: {},
    options: { default: void 0 },
    size: { default: () => getDefault("select.size") },
    variant: { default: () => getDefault("select.variant") },
    placeholder: { default: void 0 },
    disabled: { type: Boolean, default: false },
    required: { type: Boolean, default: false },
    expanded: { type: Boolean, default: false },
    rounded: { type: Boolean, default: false },
    nativeSize: { default: void 0 },
    iconPack: { default: () => getDefault("select.iconPack") },
    icon: { default: () => getDefault("select.icon") },
    iconClickable: { type: Boolean, default: false },
    iconRight: { default: () => getDefault("select.iconRight") },
    iconRightClickable: { type: Boolean, default: false },
    iconRightVariant: { default: void 0 },
    id: { default: () => useId() },
    autocomplete: { default: () => getDefault("select.autocomplete", "off") },
    statusIcon: { type: Boolean, default: () => getDefault("statusIcon", true) },
    useHtml5Validation: { type: Boolean, default: () => getDefault("useHtml5Validation", true) },
    customValidity: {},
    rootClass: {},
    sizeClass: {},
    variantClass: {},
    expandedClass: {},
    disabledClass: {},
    roundedClass: {},
    hasIconRightClass: {},
    multipleClass: {},
    selectClass: {},
    placeholderClass: {},
    arrowedClass: {},
    iconLeftSpaceClass: {},
    iconRightSpaceClass: {},
    iconLeftClass: {},
    iconRightClass: {}
  }, {
    "modelValue": {
      default: void 0
    },
    "modelModifiers": {}
  }),
  emits: /* @__PURE__ */ mergeModels(["update:model-value", "focus", "blur", "invalid", "icon-click", "icon-right-click"], ["update:modelValue"]),
  setup(__props, { expose: __expose, emit: __emit }) {
    var _a;
    const props = __props;
    const emits = __emit;
    const selectRef = useTemplateRef("selectElement");
    const { checkHtml5Validity, onBlur, onFocus, onInvalid, setFocus, isValid } = useInputHandler(selectRef, emits, props);
    const { parentField, statusVariant, statusVariantIcon } = injectField();
    if (props.id) (_a = parentField == null ? void 0 : parentField.value) == null ? void 0 : _a.setInputId(props.id);
    const vmodel = useModel(__props, "modelValue", {
      get: (value) => typeof value !== "undefined" ? value : props.multiple ? [] : "",
      set: (value) => typeof value !== "undefined" ? value : props.multiple ? [] : void 0
    });
    watch(
      vmodel,
      (value) => {
        if (parentField == null ? void 0 : parentField.value) parentField.value.setFilled(!!value);
        if (!isValid.value) checkHtml5Validity();
      },
      { immediate: true, flush: "post" }
    );
    const { nextSequence } = useSequentialId();
    const normalizedOptions = computed(
      () => normalizeOptions(props.options, nextSequence)
    );
    const placeholderVisible = computed(
      () => !isTrueish(props.multiple) && (!isDefined(vmodel.value) || vmodel.value === "")
    );
    const hasIconRight = computed(
      () => !!props.iconRight && !isTrueish(props.multiple) || props.statusIcon && !!statusVariantIcon.value
    );
    const rightIcon = computed(
      () => props.iconRight ? props.iconRight : statusVariantIcon.value
    );
    const rightIconVariant = computed(
      () => props.iconRight ? props.iconRightVariant || props.variant : statusVariant.value
    );
    function iconClick(emit, event) {
      emits(emit, event);
      nextTick(() => setFocus());
    }
    function leftIconClick(event) {
      if (props.iconClickable) iconClick("icon-click", event);
    }
    function rightIconClick(event) {
      if (props.iconRightClickable) iconClick("icon-right-click", event);
    }
    const attrs = useAttrs();
    const inputBind = computed(() => {
      var _a2;
      return {
        ...(_a2 = parentField == null ? void 0 : parentField.value) == null ? void 0 : _a2.inputAttrs,
        ...attrs
      };
    });
    const rootClasses = defineClasses(
      ["rootClass", "o-select"],
      [
        "sizeClass",
        "o-select--",
        computed(() => props.size),
        computed(() => !!props.size)
      ],
      [
        "variantClass",
        "o-select--",
        computed(() => statusVariant.value || props.variant),
        computed(() => !!statusVariant.value || !!props.variant)
      ],
      [
        "expandedClass",
        "o-select--expanded",
        null,
        computed(() => props.expanded)
      ],
      [
        "disabledClass",
        "o-select--disabled",
        null,
        computed(() => isTrueish(props.disabled))
      ],
      [
        "roundedClass",
        "o-select--rounded",
        null,
        computed(() => isTrueish(props.rounded))
      ],
      ["hasIconRightClass", "o-select--icon-right", null, hasIconRight],
      [
        "multipleClass",
        "o-select--multiple",
        null,
        computed(() => isTrueish(props.multiple))
      ]
    );
    const selectClasses = defineClasses(
      ["selectClass", "o-select__input"],
      [
        "iconLeftSpaceClass",
        "o-select__input--iconspace-left",
        null,
        computed(() => !!props.icon)
      ],
      [
        "iconRightSpaceClass",
        "o-select__input--iconspace-right",
        null,
        hasIconRight
      ],
      [
        "placeholderClass",
        "o-select__input--placeholder",
        null,
        placeholderVisible
      ],
      [
        "arrowedClass",
        "o-select__input--arrowed",
        null,
        computed(() => !hasIconRight.value && !isTrueish(props.multiple))
      ]
    );
    const iconLeftClasses = defineClasses(["iconLeftClass", "o-select__icon-left"]);
    const iconRightClasses = defineClasses([
      "iconRightClass",
      "o-select__icon-right"
    ]);
    __expose({ focus: setFocus, value: vmodel });
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock("div", {
        "data-oruga": "select",
        class: normalizeClass(unref(rootClasses))
      }, [
        _ctx.icon ? (openBlock(), createBlock(_sfc_main$1, {
          key: 0,
          class: normalizeClass(unref(iconLeftClasses)),
          clickable: _ctx.iconClickable,
          icon: _ctx.icon,
          pack: _ctx.iconPack,
          size: _ctx.size,
          onClick: _cache[0] || (_cache[0] = ($event) => leftIconClick($event))
        }, null, 8, ["class", "clickable", "icon", "pack", "size"])) : createCommentVNode("", true),
        withDirectives(createElementVNode("select", mergeProps(inputBind.value, {
          id: _ctx.id,
          ref: "selectElement",
          "onUpdate:modelValue": _cache[1] || (_cache[1] = ($event) => vmodel.value = $event),
          "data-oruga-input": "select",
          class: unref(selectClasses),
          autocomplete: _ctx.autocomplete,
          multiple: props.multiple,
          size: _ctx.nativeSize,
          disabled: _ctx.disabled,
          required: _ctx.required,
          onBlur: _cache[2] || (_cache[2] = //@ts-ignore
          (...args) => unref(onBlur) && unref(onBlur)(...args)),
          onFocus: _cache[3] || (_cache[3] = //@ts-ignore
          (...args) => unref(onFocus) && unref(onFocus)(...args)),
          onInvalid: _cache[4] || (_cache[4] = //@ts-ignore
          (...args) => unref(onInvalid) && unref(onInvalid)(...args))
        }), [
          _ctx.placeholder || _ctx.$slots.placeholder ? (openBlock(), createElementBlock(Fragment, { key: 0 }, [
            placeholderVisible.value ? (openBlock(), createElementBlock("option", _hoisted_2, [
              renderSlot(_ctx.$slots, "placeholder", {}, () => [
                createTextVNode(toDisplayString(_ctx.placeholder), 1)
              ])
            ])) : createCommentVNode("", true)
          ], 64)) : createCommentVNode("", true),
          renderSlot(_ctx.$slots, "default", {}, () => [
            (openBlock(true), createElementBlock(Fragment, null, renderList(normalizedOptions.value, (option) => {
              return openBlock(), createElementBlock(Fragment, {
                key: option.key
              }, [
                unref(isGroupOption)(option) ? withDirectives((openBlock(), createElementBlock("optgroup", mergeProps({
                  key: 0,
                  ref_for: true
                }, option.attrs, {
                  label: option.label,
                  value: option.value
                }), [
                  (openBlock(true), createElementBlock(Fragment, null, renderList(option.options, (_option) => {
                    return withDirectives((openBlock(), createElementBlock("option", mergeProps({ ref_for: true }, _option.attrs, {
                      key: _option.key,
                      value: _option.value,
                      selected: option.value === vmodel.value
                    }), toDisplayString(_option.label), 17, _hoisted_4)), [
                      [vShow, !_option.hidden]
                    ]);
                  }), 128))
                ], 16, _hoisted_3)), [
                  [vShow, !option.hidden]
                ]) : withDirectives((openBlock(), createElementBlock("option", mergeProps({
                  key: 1,
                  ref_for: true
                }, option.attrs, {
                  value: option.value,
                  selected: option.value === vmodel.value
                }), toDisplayString(option.label), 17, _hoisted_5)), [
                  [vShow, !option.hidden]
                ])
              ], 64);
            }), 128))
          ])
        ], 16, _hoisted_1), [
          [vModelSelect, vmodel.value]
        ]),
        hasIconRight.value ? (openBlock(), createBlock(_sfc_main$1, {
          key: 1,
          class: normalizeClass(unref(iconRightClasses)),
          clickable: _ctx.iconRightClickable,
          icon: rightIcon.value,
          pack: _ctx.iconPack,
          size: _ctx.size,
          variant: rightIconVariant.value,
          both: "",
          onClick: rightIconClick
        }, null, 8, ["class", "clickable", "icon", "pack", "size", "variant"])) : createCommentVNode("", true)
      ], 2);
    };
  }
});
export {
  _sfc_main as _
};
