// @oruga-ui/oruga-next/dist/cjs/Button.vue_vue_type_script_setup_true_lang-BHJX6RoE.cjs
"use strict";
/*! Oruga v0.10.5 | MIT License | github.com/oruga-ui/oruga */
const vue = require("vue");
const Icon_vue_vue_type_script_setup_true_lang = require("./Icon.vue_vue_type_script_setup_true_lang-iqIFDXno.cjs");
const config = require("./config-DQse3iF9.cjs");
const defineClasses = require("./defineClasses-HOBoThd5.cjs");
const _sfc_main = /* @__PURE__ */ vue.defineComponent({
  ...{
    isOruga: true,
    name: "OButton",
    configField: "button"
  },
  __name: "Button",
  props: {
    override: { type: Boolean, default: void 0 },
    tag: { default: () => config.getDefault("button.tag", "button") },
    variant: { default: () => config.getDefault("button.variant") },
    size: { default: () => config.getDefault("button.size") },
    label: { default: void 0 },
    iconPack: { default: () => config.getDefault("button.iconPack") },
    iconLeft: { default: void 0 },
    iconRight: { default: void 0 },
    rounded: { type: Boolean, default: () => config.getDefault("button.rounded", false) },
    expanded: { type: Boolean, default: false },
    disabled: { type: Boolean, default: false },
    outlined: { type: Boolean, default: false },
    loading: { type: Boolean, default: false },
    inverted: { type: Boolean, default: false },
    type: { default: "button" },
    iconBoth: { type: Boolean, default: false },
    rootClass: {},
    sizeClass: {},
    variantClass: {},
    outlinedClass: {},
    invertedClass: {},
    loadingClass: {},
    expandedClass: {},
    roundedClass: {},
    disabledClass: {},
    wrapperClass: {},
    iconClass: {},
    iconLeftClass: {},
    iconRightClass: {},
    labelClass: {}
  },
  emits: ["click"],
  setup(__props) {
    const props = __props;
    const computedTag = vue.computed(
      () => typeof props.disabled !== "undefined" && props.disabled !== false ? "button" : props.tag
    );
    const computedNativeType = vue.computed(
      () => props.tag === "button" || props.tag === "input" ? props.type : null
    );
    const rootClasses = defineClasses.defineClasses(
      ["rootClass", "o-button"],
      [
        "sizeClass",
        "o-button--",
        vue.computed(() => props.size),
        vue.computed(() => !!props.size)
      ],
      [
        "variantClass",
        "o-button--",
        vue.computed(() => props.variant),
        vue.computed(() => !!props.variant && !props.outlined && !props.inverted)
      ],
      [
        "outlinedClass",
        "o-button--outlined",
        null,
        vue.computed(() => props.outlined && !props.variant)
      ],
      [
        "invertedClass",
        "o-button--inverted",
        null,
        vue.computed(() => props.inverted && !props.variant)
      ],
      [
        "outlinedClass",
        "o-button--outlined-",
        vue.computed(() => props.variant),
        vue.computed(() => props.outlined && !!props.variant)
      ],
      [
        "invertedClass",
        "o-button--inverted-",
        vue.computed(() => props.variant),
        vue.computed(() => props.inverted && !!props.variant)
      ],
      [
        "expandedClass",
        "o-button--expanded",
        null,
        vue.computed(() => props.expanded)
      ],
      ["loadingClass", "o-button--loading", null, vue.computed(() => props.loading)],
      ["roundedClass", "o-button--rounded", null, vue.computed(() => props.rounded)],
      [
        "disabledClass",
        "o-button--disabled",
        null,
        vue.computed(() => props.disabled)
      ]
    );
    const wrapperClasses = defineClasses.defineClasses(["wrapperClass", "o-button__wrapper"]);
    const labelClasses = defineClasses.defineClasses(["labelClass", "o-button__label"]);
    const iconClasses = defineClasses.defineClasses(["iconClass", "o-button__icon"]);
    const iconLeftClasses = defineClasses.defineClasses(["iconLeftClass", "o-button__icon-left"]);
    const iconRightClasses = defineClasses.defineClasses([
      "iconRightClass",
      "o-button__icon-right"
    ]);
    return (_ctx, _cache) => {
      return vue.openBlock(), vue.createBlock(vue.resolveDynamicComponent(computedTag.value), {
        "data-oruga": "button",
        type: computedNativeType.value,
        role: "button",
        tabindex: "0",
        class: vue.normalizeClass(vue.unref(rootClasses)),
        disabled: _ctx.disabled ? true : null,
        onClick: _cache[0] || (_cache[0] = ($event) => _ctx.$emit("click", $event)),
        onKeydown: [
          _cache[1] || (_cache[1] = vue.withKeys(($event) => _ctx.$emit("click", $event), ["enter"])),
          _cache[2] || (_cache[2] = vue.withKeys(($event) => _ctx.$emit("click", $event), ["space"]))
        ]
      }, {
        default: vue.withCtx(() => [
          vue.createElementVNode("span", {
            class: vue.normalizeClass(vue.unref(wrapperClasses))
          }, [
            _ctx.iconLeft ? (vue.openBlock(), vue.createBlock(Icon_vue_vue_type_script_setup_true_lang._sfc_main, {
              key: 0,
              pack: _ctx.iconPack,
              icon: _ctx.iconLeft,
              size: _ctx.size,
              both: _ctx.iconBoth,
              class: vue.normalizeClass([...vue.unref(iconClasses), ...vue.unref(iconLeftClasses)])
            }, null, 8, ["pack", "icon", "size", "both", "class"])) : vue.createCommentVNode("", true),
            _ctx.label || _ctx.$slots.default ? (vue.openBlock(), vue.createElementBlock("span", {
              key: 1,
              class: vue.normalizeClass(vue.unref(labelClasses))
            }, [
              vue.renderSlot(_ctx.$slots, "default", {}, () => [
                vue.createTextVNode(vue.toDisplayString(_ctx.label), 1)
              ])
            ], 2)) : vue.createCommentVNode("", true),
            _ctx.iconRight ? (vue.openBlock(), vue.createBlock(Icon_vue_vue_type_script_setup_true_lang._sfc_main, {
              key: 2,
              pack: _ctx.iconPack,
              icon: _ctx.iconRight,
              size: _ctx.size,
              both: _ctx.iconBoth,
              class: vue.normalizeClass([...vue.unref(iconClasses), ...vue.unref(iconRightClasses)])
            }, null, 8, ["pack", "icon", "size", "both", "class"])) : vue.createCommentVNode("", true)
          ], 2)
        ]),
        _: 3
      }, 40, ["type", "class", "disabled"]);
    };
  }
});
exports._sfc_main = _sfc_main;
