// @oruga-ui/oruga-next/dist/cjs/Checkbox.vue_vue_type_script_setup_true_lang-hB1JGU94.cjs
"use strict";
/*! Oruga v0.10.5 | MIT License | github.com/oruga-ui/oruga */
const vue = require("vue");
const config = require("./config-DQse3iF9.cjs");
const defineClasses = require("./defineClasses-HOBoThd5.cjs");
const useInputHandler = require("./useInputHandler-DOYzVgrk.cjs");
const fieldInjection = require("./fieldInjection-CvE2Bm-L.cjs");
const _hoisted_1 = ["id", "name", "value", "true-value", "false-value", "required", "indeterminate", "disabled", "autocomplete", "aria-checked", "aria-labelledby"];
const _hoisted_2 = ["id", "for"];
const _sfc_main = /* @__PURE__ */ vue.defineComponent({
  ...{
    isOruga: true,
    name: "OCheckbox",
    configField: "checkbox",
    inheritAttrs: false
  },
  __name: "Checkbox",
  props: /* @__PURE__ */ vue.mergeModels({
    override: { type: Boolean, default: void 0 },
    modelValue: { default: void 0 },
    variant: { default: () => config.getDefault("checkbox.variant") },
    size: { default: () => config.getDefault("checkbox.size") },
    label: { default: void 0 },
    indeterminate: { type: Boolean, default: false },
    nativeValue: { default: void 0 },
    disabled: { type: Boolean, default: false },
    required: { type: Boolean, default: false },
    name: { default: void 0 },
    trueValue: { default: void 0 },
    falseValue: { default: void 0 },
    autocomplete: { default: () => config.getDefault("checkbox.autocomplete", "off") },
    id: { default: () => vue.useId() },
    useHtml5Validation: { type: Boolean, default: () => config.getDefault("useHtml5Validation", true) },
    customValidity: { type: [String, Function], default: "" },
    rootClass: {},
    sizeClass: {},
    variantClass: {},
    disabledClass: {},
    checkedClass: {},
    indeterminateClass: {},
    inputClass: {},
    labelClass: {}
  }, {
    "modelValue": { default: void 0 },
    "modelModifiers": {}
  }),
  emits: /* @__PURE__ */ vue.mergeModels(["update:model-value", "input", "focus", "blur", "invalid"], ["update:modelValue"]),
  setup(__props, { expose: __expose, emit: __emit }) {
    var _a, _b;
    const props = __props;
    const emits = __emit;
    const inputRef = vue.useTemplateRef("inputElement");
    const { onBlur, onFocus, onInvalid, setFocus } = useInputHandler.useInputHandler(
      inputRef,
      emits,
      props
    );
    const { parentField } = fieldInjection.injectField();
    const labelId = !!parentField.value || !!props.label || !!vue.useSlots().default ? ((_a = parentField.value) == null ? void 0 : _a.labelId) || vue.useId() : void 0;
    if (!props.label && props.id) (_b = parentField.value) == null ? void 0 : _b.setInputId(props.id);
    const vmodel = vue.useModel(__props, "modelValue");
    const isChecked = vue.computed(
      () => vmodel.value === (props.trueValue ?? true) || Array.isArray(vmodel.value) && vmodel.value.includes(props.nativeValue)
    );
    function onInput(event) {
      emits("input", vmodel.value, event);
    }
    const attrs = vue.useAttrs();
    const inputBind = vue.computed(() => {
      var _a2;
      return {
        ...(_a2 = parentField.value) == null ? void 0 : _a2.inputAttrs,
        ...attrs
      };
    });
    const rootClasses = defineClasses.defineClasses(
      ["rootClass", "o-checkbox"],
      [
        "sizeClass",
        "o-checkbox--",
        vue.computed(() => props.size),
        vue.computed(() => !!props.size)
      ],
      [
        "variantClass",
        "o-checkbox--",
        vue.computed(() => props.variant),
        vue.computed(() => !!props.variant)
      ],
      [
        "disabledClass",
        "o-checkbox--disabled",
        null,
        vue.computed(() => props.disabled)
      ],
      ["checkedClass", "o-checkbox--checked", null, isChecked],
      [
        "indeterminateClass",
        "o-checkbox--indeterminate",
        null,
        vue.computed(() => props.indeterminate)
      ]
    );
    const inputClasses = defineClasses.defineClasses(["inputClass", "o-checkbox__input"]);
    const labelClasses = defineClasses.defineClasses(["labelClass", "o-checkbox__label"]);
    __expose({ focus: setFocus, value: vmodel });
    return (_ctx, _cache) => {
      return vue.openBlock(), vue.createElementBlock("div", {
        "data-oruga": "checkbox",
        class: vue.normalizeClass(vue.unref(rootClasses))
      }, [
        vue.withDirectives(vue.createElementVNode("input", vue.mergeProps(inputBind.value, {
          id: _ctx.id,
          ref: "inputElement",
          "onUpdate:modelValue": _cache[0] || (_cache[0] = ($event) => vmodel.value = $event),
          type: "checkbox",
          "data-oruga-input": "checkbox",
          class: vue.unref(inputClasses),
          name: _ctx.name,
          value: _ctx.nativeValue,
          "true-value": _ctx.trueValue ?? true,
          "false-value": _ctx.falseValue ?? false,
          required: _ctx.required,
          indeterminate: _ctx.indeterminate,
          disabled: _ctx.disabled,
          autocomplete: _ctx.autocomplete,
          "aria-checked": _ctx.indeterminate ? "mixed" : isChecked.value,
          "aria-labelledby": vue.unref(labelId),
          onBlur: _cache[1] || (_cache[1] = //@ts-ignore
          (...args) => vue.unref(onBlur) && vue.unref(onBlur)(...args)),
          onFocus: _cache[2] || (_cache[2] = //@ts-ignore
          (...args) => vue.unref(onFocus) && vue.unref(onFocus)(...args)),
          onInvalid: _cache[3] || (_cache[3] = //@ts-ignore
          (...args) => vue.unref(onInvalid) && vue.unref(onInvalid)(...args)),
          onChange: onInput
        }), null, 16, _hoisted_1), [
          [vue.vModelCheckbox, vmodel.value]
        ]),
        _ctx.label || _ctx.$slots.default ? (vue.openBlock(), vue.createElementBlock("label", {
          key: 0,
          id: vue.unref(labelId),
          for: _ctx.id,
          class: vue.normalizeClass(vue.unref(labelClasses))
        }, [
          vue.renderSlot(_ctx.$slots, "default", {}, () => [
            vue.createTextVNode(vue.toDisplayString(_ctx.label), 1)
          ])
        ], 10, _hoisted_2)) : vue.createCommentVNode("", true)
      ], 2);
    };
  }
});
exports._sfc_main = _sfc_main;
