// @oruga-ui/oruga-next/dist/cjs/Input.vue_vue_type_script_setup_true_lang-BV9WM_5W.cjs
"use strict";
/*! Oruga v0.10.5 | MIT License | github.com/oruga-ui/oruga */
const vue = require("vue");
const Icon_vue_vue_type_script_setup_true_lang = require("./Icon.vue_vue_type_script_setup_true_lang-iqIFDXno.cjs");
const config = require("./config-DQse3iF9.cjs");
const helpers = require("./helpers.cjs");
const defineClasses = require("./defineClasses-HOBoThd5.cjs");
const useInputHandler = require("./useInputHandler-DOYzVgrk.cjs");
const useDebounce = require("./useDebounce-D4C3sk-r.cjs");
const fieldInjection = require("./fieldInjection-CvE2Bm-L.cjs");
const _hoisted_1 = ["id", "type", "data-oruga-input", "maxlength", "autocomplete", "placeholder", "disabled"];
const _hoisted_2 = ["id", "maxlength", "placeholder", "disabled"];
const _sfc_main = /* @__PURE__ */ vue.defineComponent({
  ...{
    isOruga: true,
    name: "OInput",
    configField: "input",
    inheritAttrs: false
  },
  __name: "Input",
  props: /* @__PURE__ */ vue.mergeModels({
    override: { type: Boolean, default: void 0 },
    modelValue: { default: void 0 },
    number: {},
    type: { default: "text" },
    size: { default: () => config.getDefault("input.size") },
    variant: { default: () => config.getDefault("input.variant") },
    placeholder: { default: void 0 },
    expanded: { type: Boolean, default: false },
    rounded: { type: Boolean, default: false },
    disabled: { type: Boolean, default: false },
    passwordReveal: { type: Boolean, default: false },
    maxlength: { default: void 0 },
    counter: { type: Boolean, default: () => config.getDefault("input.counter", false) },
    autosize: { type: Boolean, default: false },
    iconPack: { default: () => config.getDefault("input.iconPack") },
    icon: { default: () => config.getDefault("input.icon") },
    iconClickable: { type: Boolean, default: false },
    iconRight: { default: () => config.getDefault("input.iconRight") },
    iconRightClickable: { type: Boolean, default: false },
    iconRightVariant: { default: void 0 },
    clearable: { type: Boolean, default: () => config.getDefault("input.clearable", false) },
    clearIcon: { default: () => config.getDefault("input.clearIcon", "close-circle") },
    statusIcon: { type: Boolean, default: () => config.getDefault("statusIcon", true) },
    debounce: { default: () => config.getDefault("autocomplete.debounce", 400) },
    autocomplete: { default: () => config.getDefault("input.autocomplete", "off") },
    id: { default: () => vue.useId() },
    useHtml5Validation: { type: Boolean, default: () => config.getDefault("useHtml5Validation", true) },
    customValidity: { type: [String, Function], default: "" },
    rootClass: {},
    sizeClass: {},
    variantClass: {},
    expandedClass: {},
    roundedClass: {},
    disabledClass: {},
    textareaClass: {},
    hasIconRightClass: {},
    inputClass: {},
    iconLeftSpaceClass: {},
    iconRightSpaceClass: {},
    placeholderClass: {},
    iconLeftClass: {},
    iconRightClass: {},
    counterClass: {}
  }, {
    "modelValue": {
      // cast incomming value to string
      default: void 0
    },
    "modelModifiers": {}
  }),
  emits: /* @__PURE__ */ vue.mergeModels(["update:model-value", "input", "focus", "blur", "invalid", "icon-click", "icon-right-click"], ["update:modelValue"]),
  setup(__props, { expose: __expose, emit: __emit }) {
    var _a;
    const props = __props;
    const emits = __emit;
    const inputRef = vue.useTemplateRef("inputElement");
    const {
      checkHtml5Validity,
      onBlur,
      onFocus,
      onInvalid,
      setFocus,
      isValid,
      isFocused
    } = useInputHandler.useInputHandler(inputRef, emits, props);
    const { parentField, statusVariant, statusVariantIcon } = fieldInjection.injectField();
    const vmodel = vue.useModel(__props, "modelValue", {
      // cast incomming value to string
      get: (value) => helpers.isDefined(value) ? String(value) : "",
      // cast outgoing value to number if prop number is true
      set: (value) => !helpers.isDefined(value) ? value : helpers.isTrueish(props.number) ? Number(value) : String(value)
    });
    if (props.id) (_a = parentField == null ? void 0 : parentField.value) == null ? void 0 : _a.setInputId(props.id);
    const valueLength = vue.computed(
      () => typeof vmodel.value === "string" || typeof vmodel.value === "number" ? String(vmodel.value).length : 0
    );
    vue.onMounted(() => {
      vue.watch(
        () => vmodel.value,
        (value) => {
          if (parentField == null ? void 0 : parentField.value) parentField.value.setFilled(!!value);
          if (props.autosize) resize();
          if (!isValid.value) checkHtml5Validity();
        },
        { immediate: true, flush: "post" }
      );
    });
    const height = vue.ref("auto");
    function resize() {
      height.value = "auto";
      vue.nextTick(() => {
        if (props.type !== "textarea" || !inputRef.value) return;
        const scrollHeight = inputRef.value.scrollHeight;
        height.value = scrollHeight + "px";
      });
    }
    const computedStyles = vue.computed(
      () => props.type === "textarea" && props.autosize ? {
        resize: "none",
        height: height.value,
        overflow: "hidden"
      } : {}
    );
    let debouncedInput;
    vue.watch(
      () => props.debounce,
      (debounce) => debouncedInput = useDebounce.useDebounce(onInput, debounce || 0),
      { immediate: true }
    );
    function onInput(event) {
      emits("input", vmodel.value, event);
    }
    const placeholderVisible = vue.computed(
      () => !helpers.isDefined(vmodel.value) || vmodel.value === ""
    );
    const hasIconRight = vue.computed(() => {
      return !!(props.passwordReveal || props.statusIcon && statusVariantIcon.value || props.clearable && vmodel.value && props.clearIcon || props.iconRight);
    });
    const computedIconRight = vue.computed(() => {
      if (props.passwordReveal) return passwordVisibleIcon.value;
      else if (props.clearable && vmodel.value && props.clearIcon)
        return props.clearIcon;
      else if (props.iconRight) return props.iconRight;
      return statusVariantIcon.value;
    });
    const computedIconRightVariant = vue.computed(
      () => props.passwordReveal || props.iconRight ? props.iconRightVariant || props.variant : statusVariant.value
    );
    function iconClick(event) {
      emits("icon-click", event);
      vue.nextTick(() => setFocus());
    }
    function rightIconClick(event) {
      if (props.passwordReveal) togglePasswordVisibility();
      else if (props.clearable)
        vmodel.value = helpers.isTrueish(props.number) ? 0 : "";
      if (props.iconRightClickable) {
        emits("icon-right-click", event);
        vue.nextTick(() => setFocus());
      }
    }
    const isPasswordVisible = vue.ref(false);
    const inputType = vue.computed(() => {
      if (props.passwordReveal)
        return isPasswordVisible.value ? "text" : "password";
      else return props.type;
    });
    const passwordVisibleIcon = vue.computed(
      () => !isPasswordVisible.value ? "eye" : "eye-off"
    );
    function togglePasswordVisibility() {
      isPasswordVisible.value = !isPasswordVisible.value;
      vue.nextTick(() => setFocus());
    }
    const attrs = vue.useAttrs();
    const inputBind = vue.computed(() => {
      var _a2;
      return {
        ...(_a2 = parentField == null ? void 0 : parentField.value) == null ? void 0 : _a2.inputAttrs,
        ...attrs
      };
    });
    const rootClasses = defineClasses.defineClasses(
      ["rootClass", "o-input"],
      [
        "sizeClass",
        "o-input--",
        vue.computed(() => props.size),
        vue.computed(() => !!props.size)
      ],
      [
        "variantClass",
        "o-input--",
        vue.computed(() => statusVariant.value || props.variant),
        vue.computed(() => !!statusVariant.value || !!props.variant)
      ],
      [
        "expandedClass",
        "o-input--expanded",
        null,
        vue.computed(() => props.expanded)
      ],
      [
        "disabledClass",
        "o-input--disabled",
        null,
        vue.computed(() => props.disabled)
      ],
      ["roundedClass", "o-input--rounded", null, vue.computed(() => props.rounded)],
      ["hasIconRightClass", "o-input--icon-right", null, hasIconRight],
      [
        "textareaClass",
        "o-input--textarea",
        null,
        vue.computed(() => props.type === "textarea")
      ]
    );
    const inputClasses = defineClasses.defineClasses(
      ["inputClass", "o-input__input"],
      [
        "iconLeftSpaceClass",
        "o-input__input--iconspace-left",
        null,
        vue.computed(() => !!props.icon)
      ],
      [
        "iconRightSpaceClass",
        "o-input__input--iconspace-right",
        null,
        hasIconRight
      ],
      [
        "placeholderClass",
        "o-input__input--placeholder",
        null,
        placeholderVisible
      ]
    );
    const iconLeftClasses = defineClasses.defineClasses(["iconLeftClass", "o-input__icon-left"]);
    const iconRightClasses = defineClasses.defineClasses([
      "iconRightClass",
      "o-input__icon-right"
    ]);
    const counterClasses = defineClasses.defineClasses(["counterClass", "o-input__counter"]);
    __expose({ focus: setFocus, value: vmodel });
    return (_ctx, _cache) => {
      return vue.openBlock(), vue.createElementBlock("div", {
        "data-oruga": "input",
        class: vue.normalizeClass(vue.unref(rootClasses))
      }, [
        _ctx.icon ? (vue.openBlock(), vue.createBlock(Icon_vue_vue_type_script_setup_true_lang._sfc_main, {
          key: 0,
          class: vue.normalizeClass(vue.unref(iconLeftClasses)),
          clickable: _ctx.iconClickable,
          icon: _ctx.icon,
          pack: _ctx.iconPack,
          size: _ctx.size,
          onClick: iconClick
        }, null, 8, ["class", "clickable", "icon", "pack", "size"])) : vue.createCommentVNode("", true),
        _ctx.type !== "textarea" ? vue.withDirectives((vue.openBlock(), vue.createElementBlock("input", vue.mergeProps({ key: 1 }, inputBind.value, {
          id: _ctx.id,
          ref: "inputElement",
          "onUpdate:modelValue": _cache[0] || (_cache[0] = ($event) => vmodel.value = $event),
          type: inputType.value,
          "data-oruga-input": inputType.value,
          class: vue.unref(inputClasses),
          maxlength: _ctx.maxlength,
          autocomplete: _ctx.autocomplete,
          placeholder: _ctx.placeholder,
          disabled: _ctx.disabled,
          onBlur: _cache[1] || (_cache[1] = //@ts-ignore
          (...args) => vue.unref(onBlur) && vue.unref(onBlur)(...args)),
          onFocus: _cache[2] || (_cache[2] = //@ts-ignore
          (...args) => vue.unref(onFocus) && vue.unref(onFocus)(...args)),
          onInvalid: _cache[3] || (_cache[3] = //@ts-ignore
          (...args) => vue.unref(onInvalid) && vue.unref(onInvalid)(...args)),
          onInput: _cache[4] || (_cache[4] = //@ts-ignore
          (...args) => vue.unref(debouncedInput) && vue.unref(debouncedInput)(...args))
        }), null, 16, _hoisted_1)), [
          [vue.vModelDynamic, vmodel.value]
        ]) : vue.withDirectives((vue.openBlock(), vue.createElementBlock("textarea", vue.mergeProps({ key: 2 }, inputBind.value, {
          id: _ctx.id,
          ref: "inputElement",
          "onUpdate:modelValue": _cache[5] || (_cache[5] = ($event) => vmodel.value = $event),
          "data-oruga-input": "textarea",
          class: vue.unref(inputClasses),
          maxlength: _ctx.maxlength,
          style: computedStyles.value,
          placeholder: _ctx.placeholder,
          disabled: _ctx.disabled,
          onBlur: _cache[6] || (_cache[6] = //@ts-ignore
          (...args) => vue.unref(onBlur) && vue.unref(onBlur)(...args)),
          onFocus: _cache[7] || (_cache[7] = //@ts-ignore
          (...args) => vue.unref(onFocus) && vue.unref(onFocus)(...args)),
          onInvalid: _cache[8] || (_cache[8] = //@ts-ignore
          (...args) => vue.unref(onInvalid) && vue.unref(onInvalid)(...args)),
          onInput: _cache[9] || (_cache[9] = //@ts-ignore
          (...args) => vue.unref(debouncedInput) && vue.unref(debouncedInput)(...args))
        }), null, 16, _hoisted_2)), [
          [vue.vModelText, vmodel.value]
        ]),
        hasIconRight.value ? (vue.openBlock(), vue.createBlock(Icon_vue_vue_type_script_setup_true_lang._sfc_main, {
          key: 3,
          class: vue.normalizeClass(vue.unref(iconRightClasses)),
          clickable: _ctx.passwordReveal || _ctx.clearable || _ctx.iconRightClickable,
          icon: computedIconRight.value,
          pack: _ctx.iconPack,
          size: _ctx.size,
          variant: computedIconRightVariant.value,
          both: "",
          onClick: rightIconClick
        }, null, 8, ["class", "clickable", "icon", "pack", "size", "variant"])) : vue.createCommentVNode("", true),
        _ctx.maxlength && _ctx.counter && vue.unref(isFocused) && _ctx.type !== "number" ? (vue.openBlock(), vue.createElementBlock("small", {
          key: 4,
          class: vue.normalizeClass(vue.unref(counterClasses))
        }, vue.toDisplayString(valueLength.value) + " / " + vue.toDisplayString(_ctx.maxlength), 3)) : vue.createCommentVNode("", true)
      ], 2);
    };
  }
});
exports._sfc_main = _sfc_main;
