// @oruga-ui/oruga-next/dist/cjs/Loading.vue_vue_type_script_setup_true_lang-BcJqAqw8.cjs
"use strict";
/*! Oruga v0.10.5 | MIT License | github.com/oruga-ui/oruga */
const vue = require("vue");
const Icon_vue_vue_type_script_setup_true_lang = require("./Icon.vue_vue_type_script_setup_true_lang-iqIFDXno.cjs");
const config = require("./config-DQse3iF9.cjs");
const defineClasses = require("./defineClasses-HOBoThd5.cjs");
const useEventListener = require("./useEventListener-Cew5Aed0.cjs");
const usePreventScrolling = require("./usePreventScrolling-E0W8EUbK.cjs");
const _sfc_main = /* @__PURE__ */ vue.defineComponent({
  ...{
    isOruga: true,
    name: "OLoading",
    configField: "loading",
    inheritAttrs: false
  },
  __name: "Loading",
  props: /* @__PURE__ */ vue.mergeModels({
    override: { type: Boolean, default: void 0 },
    active: { type: Boolean, default: false },
    fullPage: { type: Boolean, default: true },
    label: { default: void 0 },
    animation: { default: () => config.getDefault("loading.animation", "fade") },
    cancelable: { type: Boolean, default: false },
    icon: { default: () => config.getDefault("loading.icon", "loading") },
    iconSpin: { type: Boolean, default: () => config.getDefault("loading.iconSpin", true) },
    iconSize: { default: () => config.getDefault("loading.iconSize", "medium") },
    clipScroll: { type: Boolean, default: () => config.getDefault("loading.clipScroll", false) },
    container: {},
    rootClass: {},
    fullPageClass: {},
    overlayClass: {},
    iconClass: {},
    labelClass: {},
    scrollClipClass: {},
    scrollKeepClass: {}
  }, {
    "fullPage": { type: Boolean, ...{ default: true } },
    "fullPageModifiers": {},
    "active": { type: Boolean, ...{ default: false } },
    "activeModifiers": {}
  }),
  emits: /* @__PURE__ */ vue.mergeModels(["update:active", "update:fullPage", "close"], ["update:fullPage", "update:active"]),
  setup(__props, { expose: __expose, emit: __emit }) {
    const props = __props;
    const emits = __emit;
    const rootRef = vue.useTemplateRef("rootElement");
    const isFullPage = vue.useModel(__props, "fullPage");
    const isActive = vue.useModel(__props, "active");
    const toggleScroll = usePreventScrolling.usePreventScrolling(props.clipScroll);
    vue.watch(isActive, (value) => {
      if (isFullPage.value) toggleScroll(value);
    });
    if (config.isClient) {
      useEventListener.useEventListener(rootRef, "keyup", onKeyPress, { trigger: isActive });
    }
    function onKeyPress(event) {
      if (!isActive.value) return;
      if (event.key === "Escape" || event.key === "Esc") cancel("escape");
    }
    function cancel(method) {
      if (!props.cancelable || Array.isArray(props.cancelable) && !props.cancelable.includes(method))
        return;
      close({ action: "cancel", method });
    }
    function close(...args) {
      isActive.value = false;
      emits("close", args);
    }
    const rootClasses = defineClasses.defineClasses(
      ["rootClass", "o-loading"],
      ["fullPageClass", "o-loading--fullpage", null, isFullPage]
    );
    const overlayClasses = defineClasses.defineClasses(["overlayClass", "o-loading__overlay"]);
    const iconClasses = defineClasses.defineClasses(["iconClass", "o-loading__icon"]);
    const labelClasses = defineClasses.defineClasses(["labelClass", "o-loading__label"]);
    __expose({ close });
    return (_ctx, _cache) => {
      return vue.openBlock(), vue.createBlock(vue.Transition, { name: _ctx.animation }, {
        default: vue.withCtx(() => [
          isActive.value ? (vue.openBlock(), vue.createElementBlock("div", {
            key: 0,
            ref: "rootElement",
            "data-oruga": "loading",
            class: vue.normalizeClass(vue.unref(rootClasses)),
            role: "status",
            "aria-live": "polite"
          }, [
            vue.createElementVNode("div", {
              class: vue.normalizeClass(vue.unref(overlayClasses)),
              tabindex: -1,
              onClick: _cache[0] || (_cache[0] = ($event) => cancel("outside"))
            }, null, 2),
            vue.renderSlot(_ctx.$slots, "default", { close }, () => [
              vue.createVNode(Icon_vue_vue_type_script_setup_true_lang._sfc_main, {
                icon: _ctx.icon,
                spin: _ctx.iconSpin,
                size: _ctx.iconSize,
                class: vue.normalizeClass(vue.unref(iconClasses)),
                both: ""
              }, null, 8, ["icon", "spin", "size", "class"]),
              _ctx.label ? (vue.openBlock(), vue.createElementBlock("span", {
                key: 0,
                class: vue.normalizeClass(vue.unref(labelClasses))
              }, vue.toDisplayString(_ctx.label), 3)) : vue.createCommentVNode("", true)
            ])
          ], 2)) : vue.createCommentVNode("", true)
        ]),
        _: 3
      }, 8, ["name"]);
    };
  }
});
exports._sfc_main = _sfc_main;
