// @oruga-ui/oruga/dist/esm/menu.mjs
import './helpers.mjs';
import './config.mjs';
import { B as BaseComponentMixin, n as normalizeComponent, b as registerComponent, u as use } from './plugins-9401a4a3.mjs';
import { _ as __vue_component__$3 } from './Icon-33c4eb50.mjs';

//
/**
 * A simple menu
 * @displayName M
 * @style _menu.scss
 */

var script = {
  name: 'OMenu',
  configField: 'menu',
  mixins: [BaseComponentMixin],
  props: {
    accordion: {
      type: Boolean,
      default: true
    },
    activable: {
      type: Boolean,
      default: true
    },
    rootClass: [String, Array, Function]
  },

  data() {
    return {
      isMenu: true // Used by MenuItem

    };
  },

  computed: {
    rootClasses() {
      return this.computedClass('rootClass', 'o-menu');
    }

  }
};

/* script */
const __vue_script__ = script;

/* template */
var __vue_render__ = function () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',{class:_vm.rootClasses},[_vm._t("default")],2)};
var __vue_staticRenderFns__ = [];

  /* style */
  const __vue_inject_styles__ = undefined;
  /* scoped */
  const __vue_scope_id__ = undefined;
  /* module identifier */
  const __vue_module_identifier__ = undefined;
  /* functional template */
  const __vue_is_functional_template__ = false;
  /* style inject */
  
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  const __vue_component__ = /*#__PURE__*/normalizeComponent(
    { render: __vue_render__, staticRenderFns: __vue_staticRenderFns__ },
    __vue_inject_styles__,
    __vue_script__,
    __vue_scope_id__,
    __vue_is_functional_template__,
    __vue_module_identifier__,
    false,
    undefined,
    undefined,
    undefined
  );

//
var script$1 = {
  name: 'OMenuList',
  configField: 'menu',
  mixins: [BaseComponentMixin],
  props: {
    ariaRole: String,
    label: String,
    icon: String,
    iconPack: String,

    /**
     * Icon size, optional
     * @values small, medium, large
     */
    size: String,
    listClass: [String, Array, Function],
    listLabelClass: [String, Array, Function]
  },
  computed: {
    listClasses() {
      return this.computedClass('listClass', 'o-menu-list');
    },

    labelClasses() {
      return this.computedClass('listLabelClass', 'o-menu-label');
    },

    computedAriaRole() {
      return this.ariaRole === 'menu' ? this.ariaRole : null;
    }

  }
};

/* script */
const __vue_script__$1 = script$1;

/* template */
var __vue_render__$1 = function () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('div',[(_vm.label || _vm.$slots.label)?_c('div',{class:_vm.labelClasses},[(_vm.label && _vm.icon)?_c('o-icon',{attrs:{"icon":_vm.icon,"pack":_vm.iconPack,"size":_vm.size}}):_vm._e(),(_vm.label)?_c('span',[_vm._v(_vm._s(_vm.label))]):_vm._t("label")],2):_vm._e(),_c('ul',{class:_vm.listClasses,attrs:{"role":_vm.computedAriaRole}},[_vm._t("default")],2)])};
var __vue_staticRenderFns__$1 = [];

  /* style */
  const __vue_inject_styles__$1 = undefined;
  /* scoped */
  const __vue_scope_id__$1 = undefined;
  /* module identifier */
  const __vue_module_identifier__$1 = undefined;
  /* functional template */
  const __vue_is_functional_template__$1 = false;
  /* style inject */
  
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  const __vue_component__$1 = /*#__PURE__*/normalizeComponent(
    { render: __vue_render__$1, staticRenderFns: __vue_staticRenderFns__$1 },
    __vue_inject_styles__$1,
    __vue_script__$1,
    __vue_scope_id__$1,
    __vue_is_functional_template__$1,
    __vue_module_identifier__$1,
    false,
    undefined,
    undefined,
    undefined
  );

//
var script$2 = {
  name: 'OMenuItem',
  components: {
    [__vue_component__$3.name]: __vue_component__$3
  },
  mixins: [BaseComponentMixin],
  configField: 'menu',
  inheritAttrs: false,
  props: {
    label: String,
    active: Boolean,
    expanded: Boolean,
    disabled: Boolean,
    iconPack: String,
    icon: String,
    animation: {
      type: String,
      default: 'slide'
    },
    tag: {
      type: String,
      default: 'a'
    },
    ariaRole: {
      type: String,
      default: ''
    },

    /**
     * Icon size, optional
     * @values small, medium, large
     */
    size: String,
    itemClass: [String, Array, Function],
    itemActiveClass: [String, Array, Function],
    itemDisabledClass: [String, Array, Function],
    itemIconTextClass: [String, Array, Function],
    itemSubmenuClass: [String, Array, Function],
    itemWrapperClass: [String, Array, Function]
  },

  data() {
    return {
      newActive: this.active,
      newExpanded: this.expanded
    };
  },

  computed: {
    ariaRoleMenu() {
      return this.ariaRole === 'menuitem' ? this.ariaRole : null;
    },

    itemClasses() {
      return {
        [this.computedClass('itemClass', 'o-menu-item')]: true,
        [this.computedClass('itemActiveClass', 'o-menu-item--active')]: this.newActive,
        [this.computedClass('itemDisabledClass', 'o-menu-item--disabled')]: this.disabled,
        [this.computedClass('itemIconTextClass', 'o-menu-item--icon-text')]: this.icon
      };
    },

    submenuClasses() {
      return this.computedClass('itemSubmenuClass', 'o-menu-item__submenu');
    },

    wrapperClasses() {
      return this.computedClass('itemWrapperClass', 'o-menu-item__wrapper');
    }

  },
  watch: {
    active(value) {
      this.newActive = value;
    },

    expanded(value) {
      this.newExpanded = value;
    }

  },
  methods: {
    onClick() {
      if (this.disabled) return;
      const menu = this.getMenu();
      this.reset(this.$parent, menu);
      this.newExpanded = this.$props.expanded || !this.newExpanded;
      this.$emit('update:expanded', this.newExpanded);

      if (menu && menu.activable) {
        this.newActive = true;
        this.$emit('update:active', this.newActive);
      }
    },

    reset(parent, menu) {
      const items = parent.$children.filter(c => c.name === this.name);
      items.forEach(item => {
        if (item !== this) {
          this.reset(item, menu);

          if (!parent.$data.isMenu || parent.$data.isMenu && parent.accordion) {
            item.newExpanded = false;
            item.$emit('update:expanded', item.newActive);
          }

          if (menu && menu.activable) {
            item.newActive = false;
            item.$emit('update:active', item.newActive);
          }
        }
      });
    },

    getMenu() {
      let parent = this.$parent;

      while (parent && !parent.$data.isMenu) {
        parent = parent.$parent;
      }

      return parent;
    }

  }
};

/* script */
const __vue_script__$2 = script$2;

/* template */
var __vue_render__$2 = function () {var _vm=this;var _h=_vm.$createElement;var _c=_vm._self._c||_h;return _c('li',{class:_vm.wrapperClasses,attrs:{"role":_vm.ariaRoleMenu}},[_c(_vm.tag,_vm._g(_vm._b({tag:"component",class:_vm.itemClasses,on:{"click":function($event){return _vm.onClick()}}},'component',_vm.$attrs,false),_vm.$listeners),[(_vm.icon)?_c('o-icon',{attrs:{"icon":_vm.icon,"pack":_vm.iconPack,"size":_vm.size}}):_vm._e(),(_vm.label)?_c('span',[_vm._v(" "+_vm._s(_vm.label)+" ")]):_vm._t("label",null,{"expanded":_vm.newExpanded,"active":_vm.newActive})],2),(_vm.$slots.default)?[_c('transition',{attrs:{"name":_vm.animation}},[_c('ul',{directives:[{name:"show",rawName:"v-show",value:(_vm.newExpanded),expression:"newExpanded"}],class:_vm.submenuClasses},[_vm._t("default")],2)])]:_vm._e()],2)};
var __vue_staticRenderFns__$2 = [];

  /* style */
  const __vue_inject_styles__$2 = undefined;
  /* scoped */
  const __vue_scope_id__$2 = undefined;
  /* module identifier */
  const __vue_module_identifier__$2 = undefined;
  /* functional template */
  const __vue_is_functional_template__$2 = false;
  /* style inject */
  
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  const __vue_component__$2 = /*#__PURE__*/normalizeComponent(
    { render: __vue_render__$2, staticRenderFns: __vue_staticRenderFns__$2 },
    __vue_inject_styles__$2,
    __vue_script__$2,
    __vue_scope_id__$2,
    __vue_is_functional_template__$2,
    __vue_module_identifier__$2,
    false,
    undefined,
    undefined,
    undefined
  );

const Plugin = {
  install(Vue) {
    registerComponent(Vue, __vue_component__);
    registerComponent(Vue, __vue_component__$1);
    registerComponent(Vue, __vue_component__$2);
  }

};
use(Plugin);

export default Plugin;
export { __vue_component__ as OMenu, __vue_component__$2 as OMenuItem, __vue_component__$1 as OMenuList };
